blockchain/state_manager/src/lib.rs

```diff
@@ -887,4 +888,42 @@ where
         ps.miner_nominal_power_meets_consensus_minimum(self.blockstore(), addr)
             .map_err(|e| e.to_string())
     }
+
+    pub async fn validate_chain<V: ProofVerifier>(
+        &self,
+        ts: Tipset,
+    ) -> Result<(), Box<dyn StdError>> {
+        let mut ts_chain: Vec<Tipset> = vec![ts.clone()];
+        let mut ts = ts;
+        while ts.epoch() != 0 {
+            let next = chain::tipset_from_keys(self.blockstore(), ts.parents())?;
+            ts_chain.push(next.clone());
+            ts = next;
+        }
+        let mut last_state = ts_chain.last().unwrap().parent_state().clone();
+        let (_, mut last_receipt) = self.tipset_state::<V>(ts_chain.last().unwrap()).await?;
```

Comment:
why not just use the receipts from the tipset as the above does? It's not exposed right now, but there is no reason it shouldn't be. I'm not confident this will always work and it's unnecessary cycles